core-backend/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Provide sp-sandbox support.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+extern crate core;
+
+pub mod env;
+mod funcs;
+pub mod memory;
+#[cfg(any(feature = "mock", test))]
+pub mod mock;
+pub mod runtime;
+mod state;
+
+use codec::{Decode, Encode};
+pub use env::Environment;
+pub use memory::{DefaultExecutorMemory, MemoryWrap};
+
+use actor_system_error::actor_system_error;
+use gear_core::{
+    env::Externalities,
+    gas::{ChargeError, CounterType, CountersOwner, GasAmount},
+    ids::ProgramId,
+    memory::MemoryInterval,
+    message::MessageWaitedType,
+    str::LimitedStr,
+};
+use gear_core_errors::ExtError as FallibleExtError;
+use gear_lazy_pages_common::ProcessAccessError;
+
+actor_system_error! {
+    pub type TerminationReason = ActorSystemError<ActorTerminationReason, SystemTerminationReason>;
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, derive_more::From)]
+pub enum UndefinedTerminationReason {
```

Comment:
don't you want to separate everything related to errors and its reason into its own module?